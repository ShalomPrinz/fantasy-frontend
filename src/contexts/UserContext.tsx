import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from "react";

import { toast } from "react-toastify";

import {
  getUserInfo,
  loadIdToken,
  loginUser,
  logoutUser,
  registerUser,
  removeIdToken,
  saveIdToken,
} from "../services";
import { LoginUser, RegisterUser, Team, User } from "../types";

enum State {
  LOADING_USER,
  LOGGED_USER,
  NO_LOGGED_USER,
}

interface UserContextValue {
  loading: boolean;
  login: (info: {}) => Promise<void>;
  logout: () => Promise<void>;
  register: (info: {}) => Promise<void>;
  state: State;
  user: User | undefined;
}

const UserContext = createContext<UserContextValue | undefined>(undefined);

function useUser() {
  const context = useContext(UserContext);
  if (typeof context === "undefined") {
    throw new Error("useUser must be used within a UserProvider");
  }
  return context;
}

interface UserProviderProps {
  children: ReactNode;
}

function UserProvider({ children }: UserProviderProps) {
  const [loading, setLoading] = useState(true);
  const [currentUser, setCurrentUser] = useState<User | undefined>();
  const state =
    typeof currentUser === "undefined"
      ? loading
        ? State.LOADING_USER
        : State.NO_LOGGED_USER
      : State.LOGGED_USER;

  const loadUser = async () => {
    if (loadIdToken()) {
      setLoading(true);
      const res = await getUserInfo().catch(({ response }) => {
        if (response.status === 401) {
          toast.error(
            "Your saved login info is outdated. Please Log In again",
            {
              toastId: "outdated",
            }
          );
          removeIdToken();
        }
      });
      if (res?.data?.user) {
        const { leagues, nickname, team } = res.data.user;
        const userTeam = new Team(team, true);
        const user = new User(leagues, nickname, userTeam);
        setCurrentUser(user);
      }
    }
    setLoading(false);
  };

  useEffect(() => {
    loadUser();
  }, []);

  const login = async (loginInfo: LoginUser) => {
    const idToken = await loginUser(loginInfo);
    saveIdToken(idToken);
    loadUser();
  };

  const logout = async () => {
    removeIdToken();
    setCurrentUser(undefined);
    await logoutUser();
  };

  const register = async (registerInfo: RegisterUser) => {
    const idToken = await registerUser(registerInfo);
    saveIdToken(idToken);
    loadUser();
  };

  const value: UserContextValue = {
    loading,
    login,
    logout,
    register,
    state,
    user: currentUser,
  };

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
}

export { State, useUser };
export default UserProvider;
