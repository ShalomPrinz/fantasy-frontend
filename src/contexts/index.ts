import TeamProvider from "./TeamContext";
import UserProvider from "./UserContext";

export { useTeamState, useTeamUpdate } from "./TeamContext";
export { useUser } from "./UserContext";
export { TeamProvider, UserProvider };
