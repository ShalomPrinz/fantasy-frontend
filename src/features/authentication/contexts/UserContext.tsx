import {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from "react";

import { toast } from "react-toastify";

import type { User } from "../../../types";
import {
  getUserInfo,
  loadIdToken,
  loginUser,
  logoutUser,
  registerUser,
  removeIdToken,
  saveIdToken,
} from "../services";
import { parseUser } from "../types";
import type { LoginUser, RegisterUser } from "../types";

enum UserState {
  LOADING_USER,
  LOGGED_USER,
  NO_LOGGED_USER,
}

interface UserContextValue {
  invalidate: () => Promise<void>;
  loading: boolean;
  login: (info: {}) => Promise<void>;
  logout: () => Promise<void>;
  register: (info: {}) => Promise<void>;
  state: UserState;
  user: User | undefined;
}

const UserContext = createContext<UserContextValue | undefined>(undefined);

function useUser() {
  const context = useContext(UserContext);
  if (typeof context === "undefined") {
    throw new Error("useUser must be used within a UserProvider");
  }
  return context;
}

interface UserProviderProps {
  children: ReactNode;
}

function UserProvider({ children }: UserProviderProps) {
  const [loading, setLoading] = useState(true);
  const [currentUser, setCurrentUser] = useState<User | undefined>();
  const state =
    typeof currentUser === "undefined"
      ? loading
        ? UserState.LOADING_USER
        : UserState.NO_LOGGED_USER
      : UserState.LOGGED_USER;

  const loadUser = async () => {
    if (loadIdToken()) {
      setLoading(true);
      const res = await getUserInfo().catch(({ response }) => {
        if (response.status === 401) {
          toast.error(
            "Your saved login info is outdated. Please Log In again",
            {
              toastId: "outdated",
            }
          );
          removeIdToken();
        }
      });
      if (res?.data?.user) {
        const user = parseUser(res.data.user);
        setCurrentUser(user);
      }
    }
    setLoading(false);
  };

  useEffect(() => {
    loadUser();
  }, []);

  const login = async (loginInfo: LoginUser) => {
    const idToken = await loginUser(loginInfo);
    saveIdToken(idToken);
    loadUser();
  };

  const logout = async () => {
    removeIdToken();
    setCurrentUser(undefined);
    await logoutUser();
  };

  const register = async (registerInfo: RegisterUser) => {
    const idToken = await registerUser(registerInfo);
    saveIdToken(idToken);
    loadUser();
  };

  const value: UserContextValue = {
    invalidate: loadUser,
    loading,
    login,
    logout,
    register,
    state,
    user: currentUser,
  };

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
}

export { UserProvider, UserState, useUser };
