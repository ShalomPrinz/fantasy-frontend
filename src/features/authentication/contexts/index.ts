export { UserProvider, UserState, useUser } from "./UserContext";
