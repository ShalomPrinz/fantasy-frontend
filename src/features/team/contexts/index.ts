import TeamProvider from "./TeamContext";

export { useTeamState, useTeamUpdate } from "./TeamContext";
export { TeamProvider };
