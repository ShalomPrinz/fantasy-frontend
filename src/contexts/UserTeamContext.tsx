import { createContext, ReactNode, useContext, useReducer } from "react";

import { Player, Team } from "../types";
import { useUser } from "./UserContext";

type Action = { type: "addPlayer"; payload: { player: Player } };
type Dispatch = (action: Action) => void;

const TeamContextState = createContext<Team | undefined>(undefined);
function useTeamState() {
  const team = useContext(TeamContextState);
  if (team === undefined)
    throw new Error("useTeamState must be used within a TeamProvider");
  return team;
}

const TeamContextUpdate = createContext<Dispatch | undefined>(undefined);
function useTeamUpdate() {
  const dispatch = useContext(TeamContextUpdate);
  if (dispatch === undefined)
    throw new Error("useTeamUpdate must be used within a TeamProvider");
  return (player: Player) =>
    dispatch({ type: "addPlayer", payload: { player } });
}

interface TeamProviderProps {
  children: ReactNode;
}

function TeamProvider({ children }: TeamProviderProps) {
  const { user } = useUser();
  const [team, dispatch] = useReducer(teamReducer, user!.team);

  return (
    <TeamContextState.Provider value={team}>
      <TeamContextUpdate.Provider value={dispatch}>
        {children}
      </TeamContextUpdate.Provider>
    </TeamContextState.Provider>
  );
}

function teamReducer(team: Team, action: Action) {
  switch (action.type) {
    case "addPlayer": {
      const { player } = action.payload;
      if (!team.contains(player)) {
        team.addPlayer(player);
      }
      return team.clone();
    }
  }
}

export default TeamProvider;
export { useTeamState, useTeamUpdate };
