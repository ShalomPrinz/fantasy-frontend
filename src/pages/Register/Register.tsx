import * as Yup from "yup";

import { AuthForm, AuthProps, Message } from "../../components";
import { UserState, useUser } from "../../contexts";

const schema = Yup.object({
  fullName: Yup.string()
    .max(30, "Must be 30 characters or less")
    .required("Required"),
  username: Yup.string()
    .max(10, "Must be 10 characters or less")
    .required("Required"),
  email: Yup.string().email("Invalid email address").required("Required"),
  password: Yup.string()
    .min(6, "Must be 6 characters or more")
    .required("Required"),
});

const textInputs = [
  {
    id: 0,
    label: "Full Name",
    name: "fullName",
    type: "text",
    placeholder: "Jane Doe",
  },
  {
    id: 1,
    label: "Username",
    name: "username",
    type: "text",
    placeholder: "Jane",
  },
  {
    id: 2,
    label: "Email",
    name: "email",
    type: "email",
    placeholder: "Jane.Doe@gmail.com",
  },
  {
    id: 3,
    label: "Password",
    name: "password",
    type: "password",
    placeholder: "******",
  },
];

const props: AuthProps = {
  other: {
    message: "If you already have a user, you should log in instead",
    title: "Log In",
    url: "/login",
  },
  title: "Register",
  toastError: "Couldn't register. Please try again later",
  toastSuccess: "Successfully registered. Enjoy!",
};

const Register = () => {
  const { register, state } = useUser();

  switch (state) {
    case UserState.LOADING_USER:
      return <Message color="info" text="Loading..." />;

    case UserState.LOGGED_USER:
      return <Message color="success" text="You are already logged in." />;

    case UserState.NO_LOGGED_USER:
      return (
        <AuthForm
          authFn={register}
          authProps={props}
          schema={schema}
          textInputs={textInputs}
        />
      );
  }
};
export default Register;
