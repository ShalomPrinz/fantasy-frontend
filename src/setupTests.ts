// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import axios from "axios";
import { setupServer } from "msw/node";

import { handlers } from "../test";
import * as TeamContext from "./contexts/TeamContext";
import * as UserContext from "./features/authentication/contexts/UserContext";
import { LeagueInfo, Message, Player, Team, User } from "./types";

jest.setTimeout(10000);

axios.defaults.baseURL = "";

const server = setupServer(...handlers);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

interface MockedUserContext {
  logout?: () => Promise<void>;
  state?: UserContext.UserState;
  user?: User;
}

interface MockedUser {
  inbox?: Message[];
  leagues?: LeagueInfo[];
  name?: string;
  players?: Player[];
}

export const getTestUser = (user?: MockedUser): User => ({
  id: "id",
  inbox: user?.inbox || [],
  leagues: user?.leagues || [],
  name: user?.name || "",
  team: new Team(user?.players || []),
});

export const mockUser = (context: MockedUserContext) => {
  const userContext = jest.spyOn(UserContext, "useUser");

  const { logout, state, user } = context;
  const mockedLogout = typeof logout !== "undefined" ? logout : async () => {};
  const mockedState =
    typeof state !== "undefined" ? state : UserContext.UserState.LOGGED_USER;
  const mockedUser = typeof user !== "undefined" ? user : getTestUser();

  userContext.mockImplementation(() => ({
    loading: false,
    login: async () => {},
    logout: mockedLogout,
    register: async () => {},
    state: mockedState,
    user: mockedUser,
    invalidate: async () => {},
  }));

  return userContext.mockRestore;
};

export const mockTeam = (team: Team) => {
  const teamContext = jest.spyOn(TeamContext, "useTeamState");
  teamContext.mockImplementation(() => team);
};

export * from "../test";
