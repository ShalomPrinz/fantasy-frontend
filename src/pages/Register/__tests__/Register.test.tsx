import { UserState } from "contexts";
import { mockUser, renderWithRouterAndUser } from "setupTests";

import Register from "../Register";

const renderComponent = () => renderWithRouterAndUser(<Register />);

describe("Register", () => {
  it("should render Register component if loading user", () => {
    mockUser({ state: UserState.LOADING_USER });
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render Register component if user is logged in", () => {
    mockUser({ state: UserState.LOGGED_USER });
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render Register component if no user is logged in", () => {
    mockUser({ state: UserState.NO_LOGGED_USER });
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });
});
