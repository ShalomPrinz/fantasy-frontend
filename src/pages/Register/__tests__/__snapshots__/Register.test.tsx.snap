// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should render Register component if loading user 1`] = `
<DocumentFragment>
  <h1
    class="text-center my-5 p-5 bg-info text-white w-75 mx-auto"
  >
    Loading...
  </h1>
</DocumentFragment>
`;

exports[`Register should render Register component if no user is logged in 1`] = `
<DocumentFragment>
  <main
    class="container"
  >
    <div
      class="m-4 row"
    >
      <div
        class="col"
      >
        <form
          action="#"
          class="container"
        >
          <div
            class="row"
          >
            <h1>
              Register
            </h1>
          </div>
          <div
            class="mx-2 px-2 py-2 row"
          >
            <label
              for="fullName"
            >
              Full Name
            </label>
            <input
              class="form-text-input"
              id="0"
              name="fullName"
              placeholder="Jane Doe"
              type="text"
              value=""
            />
          </div>
          <div
            class="mx-2 px-2 py-2 row"
          >
            <label
              for="username"
            >
              Username
            </label>
            <input
              class="form-text-input"
              id="1"
              name="username"
              placeholder="Jane"
              type="text"
              value=""
            />
          </div>
          <div
            class="mx-2 px-2 py-2 row"
          >
            <label
              for="email"
            >
              Email
            </label>
            <input
              class="form-text-input"
              id="2"
              name="email"
              placeholder="Jane.Doe@gmail.com"
              type="email"
              value=""
            />
          </div>
          <div
            class="mx-2 px-2 py-2 row"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              autocomplete="on"
              class="form-text-input"
              id="3"
              name="password"
              placeholder="******"
              type="password"
              value=""
            />
          </div>
          <div
            class="row"
          >
            <button
              class="fs-2 bg-default py-2 px-auto mt-3 rounded button-border-focus"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
      <div
        class="m-5 text-center col"
      >
        <h3
          class="p-5"
        >
          If you already have a user, you should log in instead
        </h3>
        <a
          class="fs-2 bg-default py-3 px-5 rounded text-decoration-none button-border-focus"
          href="/login"
        >
          Log In
        </a>
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`Register should render Register component if user is logged in 1`] = `
<DocumentFragment>
  <h1
    class="text-center my-5 p-5 bg-success text-white w-75 mx-auto"
  >
    You are already logged in.
  </h1>
</DocumentFragment>
`;
